### DISCUSSION TITLE
Emitting Rust code for embedded platforms

### DISCUSSION THREAD

**Discussion by pkese** [fable-compiler/Fable#3398] [General]
@ncave Do you think it would be possible to somehow instruct Fable to emit startup code and interrupt routine decorators compatible with embedded microcontroller platforms.

It feels as if programming microcontrollers could be a cool target (and driver of growth) for the Fable-Rust, as it is a small limited use-case with a fixed amount of HW support libraries that need to be covered.

I was looking at https://esp-rs.github.io/book/ and https://esp-rs.github.io/espressif-trainings/ but I don't know exactly how to trick Fable-Rust to emit compatible code and decorators.

Also something, that's beyond my knowledge is, whether compiling with fable-library-rust requires Standard Library 'std' or can one target bare metal 'no_std' as well -- https://esp-rs.github.io/book/overview/index.html

**Comment by ncave** [comment]
@pkese No-std is supported (with `no_std` feature flag on `fable-library-rust`), so it should be possible.
There is also support for inner/outer attibutes, if that is needed, like this:
```fs
[<OuterAttr("somename", [|"somekey=\"some value\""|])>]
```
Writing native bindings is a bit tricky, as you have to match the external and internal types.
Perhaps you can start with calling the F#-Rust generated code as a library, if it makes is easier.
If you have some very simple Rust example, I can take a look to see if it can be made to work.
